### PR TITLE
ipq40xx: Add support for D-Link DAP-2610

### PR DESCRIPTION
**Pull request by frolswe**
Specifications
==============
- SOC: IPQ4018
- RAM: DDR3 256MB
- Flash: SPI NOR 16MB
- WiFi:
    - 2.4GHz: IPQ4018, 2x2, front end SKY85303-11
    - 5GHz: IPQ4018, 2x2, front end SKY85717-21
- Ethernet: 1x 10/100/1000Mbps, POE 802.3af
- PHY: QCA8072
- UART: GND, blocked, 3.3V, RX, TX / 115200 8N1
- LED: 1x red / green
- Button: 1x reset / factory default
- U-Boot bootloader with tftp and "emergency web server" accessible using serial port.

Installation
============
Flash factory image from D-Link web UI. Constraints in the D-Link web UI
makes the factory image unnecessarily large. Flash again using sysupgrade
from inside OpenWrt to reclaim some flash space.

Return to stock D-Link firmware
===============================
Partition layout is preserved, and it is possible to return to the stock
firmware simply by downloading it from D-Link and writing it to the
firmware partition.

    # mtd -r write dap2610-firmware.bin firmware

Quirks
======
To be flashable from the D-Link http server, the firmware must be larger
then 6MB, and the size in the firmware header must match the actual file
size. Also, the boot loader verifies the checksum of the firmware before
each boot, thus the jffs2 must be after the checksum covered part. This
is solved in the factory image by having the rootfs at the very end of
the image (without pad-rootfs).

The sysupgrade image which does not have to be flashable from the D-Link
web UI may be smaller, and the checksum in the firmware header only
covers the kernel part of the image.

Changelog
=========
2020-01-17
----------
- Rebased to master, d4009d7.
- Move wrgg-iamge to target Makefile (and simplify).
- Set BLOCKSIZE.

2020-01-09
----------
- Rebased to master, 414d054138.
- Use pad-extra instead of pad-to in beginning of image.
- Fix broken rebase/merge.
- Rename to qcom-ipq4018-dap-2610.dts.
- Make LEDs behave more like stated in the DAP-2610 manual.
- Rename flash partition nodes to 'partition@...'.
- Add label_mac.
- Rebase dts on qcom-ipq4019.dtsi.

### CODE REVIEW

File: target/linux/ipq40xx/image/Makefile [hunk 1]
```diff
@@ -169,6 +169,31 @@ define Device/compex_wpj428
...
+	# Bootloader expects a special 160 byte header which is updated by
+	# wrgg-image.
+	# Factory image size must be larger than 6MB, and size in wrgg header must
+	# match actual factory image size to be flashable from D-Link http server.
+	# Bootloader verifies checksum of wrgg image before booting, thus jffs2
+	# cannot be part of the wrgg image. This is solved in the factory image by
+	# having the rootfs at the end of the image (without pad-rootfs). And in
+	# the sysupgrade image only the kernel is included in the wrgg checksum,
+	# but this is not flashable from the D-link http server.
+	# append-rootfs must start on an erase block boundary.
```
Comment (adrianschmutzler) [inline target/linux/ipq40xx/image/Makefile]: Consider moving that explanation to the commit message.

Comment (frolswe) [reply target/linux/ipq40xx/image/Makefile]: Sure, I will update the commit message.

**Comment by frolswe** [comment]
I missed adding the wifi board calibration files to this pull request. I will update it when that is fixed.
I don't foresee any other changes though, please comment if you want me to change anything.

File: target/linux/ipq40xx/image/Makefile (hunk 1 above)
Comment (frolswe) [reply target/linux/ipq40xx/image/Makefile]: FIxed.

**Comment by frolswe** [comment]
The current pull request is complete with the board calibration files. Please consider merging this.

File: target/linux/ipq40xx/image/Makefile (hunk 1 above)
Comment (adrianschmutzler) [reply target/linux/ipq40xx/image/Makefile]: With "moving", I implied that you remove it here afterwards, so we do not keep this long explanation in the Makefile. Whoever wonders about what's done there will be able to check the commit adding it.

Having scanned through the file, I however encountered several similar cases, so I will leave this decision to the committer reviewing this after me.

**Comment by adrianschmutzler** [comment]
> I don't foresee any other changes though, please comment if you want me to change anything.

I just commented on that particular thing, I'm not familiar enough with the target to review your commit. You will have to wait for someone else for a real review.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+#include "qcom-ipq4019-ap.dk01.1.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "D-Link DAP 2610";
+	compatible = "dlink,dap-2610";
+
+	soc {
+		edma@c080000 {
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Spacing is wrong here

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	};
+};
+
+&gmac0 {
+	vlan_tag = <2 0x20>;
+};
+
+&blsp1_spi1 {
+	flash@0 {
+		compatible = "w25q128", "jedec,spi-nor";
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: w25q128 is not needed here, "jedec,spi-nor" will auto id it.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0>;
+		spi-max-frequency = <33000000>;
```
Comment (robimarko) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Where is this from?
Other ipq40xx boards are running at 24000000

File: target/linux/ipq40xx/image/Makefile (hunk 1 above)
Comment (robimarko) [reply target/linux/ipq40xx/image/Makefile]: This is a common case, makes life really easier if some updating and reworking is done later

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 2 above)
Comment (frolswe) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: I don't understand how you want me to change it. Please explain.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 4 above)
Comment (frolswe) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: It is from the w25q128 specification. I guess it is safer to user keep 24000000 here as well if D-Link decides to change chip or whatnot.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 3 above)
Comment (frolswe) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: The included qcom-ipq4019-ap.dk01.1.dtsi contains a "compatible = mx25l25635e" for this node. So while it is not required, omitting it here will lead to a warning similar to "expected mx25..., found w25q.." when the chip is detected.  
I'm fine either way, or even removing the compatible from the included dtsi.

Comment (frolswe) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Humm, you are correct. I cannot provoke this warning anymore. I will change to only "jedec,spi-nor".  
(My understanding of how nodes in device-tree are overridden is somewhat lacking. It seems my flash@0 node replaces the mx25l25635e@0 from the dtsi file, as opposed to only overriding individual values.)

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 4 above)
Comment (robimarko) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Well, it cant runt at 33 MHz as IPQ4019 has a controller with max of 25 MHz.
That is why QSDK and everybody just uses 24 MHz to be safe.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 3 above)
Comment (robimarko) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Even if that warning occurs its not a issue, its just a warning and NOR will get properly identified.
You usually add the compatible if there is some special quirk or you have a model which does not auto ID properly.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 2 above)
Comment (robimarko) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Sorry, it looks wrong here but looking at it in a proper editor revels that its normal.

**Comment by frolswe** [comment]
Thanks for reviewing @robimarko. I pushed pushed the frequency and flash chip changes.
(Sorry for pushing the wrong commit I did not mean to rebase the pull request. It should be correct now.)

**Comment by frolswe** [comment]
Good morning.
Can I do anything more to move this pull request forward? Anything needs changing, or anyone who should be pinged?
BR
/Fredrik

**Comment by adrianschmutzler** [comment]
Please rebase onto recent master.

**Comment by frolswe** [comment]
Rebased to current master, fixing conflicts and updating to the new caldata scripts.

File: include/image.mk [hunk 5]
```diff
@@ -181,7 +181,11 @@ ifeq ($(strip $(call kernel_patchver_ge,4.18.0)),1)
 endif
 
 define Image/pad-to
-	dd if=$(1) of=$(1).new bs=$(2) conv=sync
+	if [ -e $(1) ]; then \
+		dd if=$(1) of=$(1).new bs=$(2) conv=sync; \
+	else \
+		dd if=/dev/zero of=$(1).new bs=$(2) count=1 conv=sync; \
```
Comment (chunkeey) [inline include/image.mk]: it looks to me that you could use "pad-extra" over "pad-to" in this case.

File: package/firmware/ipq-wifi/Makefile [hunk 6]
```diff
@@ -22,9 +22,13 @@ endef
 # You can find instructions and examples on the linux-wireless wiki:
 # <https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles>
 #
+# Use ath10k-bdencoder from https://github.com/qca/qca-swiss-army-knife.git
+# to generate the board-* files here.
+#
 
 ALLWIFIBOARDS:= \
 	avm_fritzrepeater-1200 \
+	dlink_dap2610 \
```
Comment (chunkeey) [inline package/firmware/ipq-wifi/Makefile]: Have you posted the boardfiles to the ath10k mailinglist as well? If not, then please follow the instructions in the comment above and post these to the mailing-list. Thanks.

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 7]
```diff
@@ -40,6 +40,11 @@ ipq40xx_setup_interfaces()
 			"0u@eth0" "1:lan" "2:lan" "3:lan" "4:lan"
 		;;
 	avm,fritzrepeater-1200|\
+	avm,fritzrepeater-3000|\
```
Comment (chunkeey) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network]: I think this is a merge gone wrong? There are now two entries in 02_network for avm,fritzrepeater-3000 and the compex,wpj428.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+	model = "D-Link DAP 2610";
+	compatible = "dlink,dap-2610";
+
+	soc {
+		edma@c080000 {
+			qcom,num_gmac = <1>;
+		};
+	};
+
+	aliases {
```
Comment (chunkeey) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: (This is optional, but I would move the aliases node in front of the soc-node)

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		led-boot = &power;
```
Comment (chunkeey) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: The [manual of the DAP-2610](ftp://ftp.dlink.de/dap/dap-2610/documentation/DAP-2610_man_reva_Manual_en.pdf) has the following to say about the LED - Indicators.

"Solid Red - Indicates the access point has malfunctioned."
"Blinking Red - This LED will blink during boot-up."
"Solid Green - Indicates   that   the   DAP-2610   is   working   properly"

Based on this description, it should be possible to mimic the same LED behavior with OpenWrt.
For example, if you add a label to the red led (i.e. led_redstatus) below and change the led-failsafe and led-boot aliases to:

led-boot = &led_redstatus;
led-failsafe = &led_redstatus;

the AP should behave like the DAP-2610 manual states (which is a plus).

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+		compatible = "jedec,spi-nor";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0>;
+		spi-max-frequency = <24000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+
+			SBL1@0 {
```
Comment (chunkeey) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Or "partition@..." .

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts [hunk 11]
```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq4019-ap.dk01.1.dtsi"
```
Comment (chunkeey) [inline target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: I would prefer if you could rebase this DTS on top of the qcom-ipq4019.dtsi and not the "qcom-ipq4019-ap.dk01.1.dtsi".
Reason being that "qcom-ipq4019-ap.dk01.1.dtsi" is part of qualcomm's EVA boards and out of our control (you could end up with unrelated changes in these files that break your device).

File: target/linux/ipq40xx/image/Makefile [hunk 12]
```diff
@@ -193,6 +193,32 @@ define Device/compex_wpj428
 endef
 TARGET_DEVICES += compex_wpj428
 
+define Device/dlink_dap-2610
+	$(call Device/FitImageLzma)
+	DEVICE_VENDOR := D-Link
+	DEVICE_MODEL := DAP-2610
+	DEVICE_DTS := qcom-ipq4018-dlink-dap-2610
```
Comment (chunkeey) [inline target/linux/ipq40xx/image/Makefile]: Due to 9c7025b7794ae06774632499547f78d827d13499 , DEVICE_DTS is dropped in favor of 

SOC := ipq4018

(you'll need to rename the dts to qcom-ipq4018-dap-2610.dts)

File: target/linux/ipq40xx/image/Makefile [hunk 13]
```diff
@@ -193,6 +193,32 @@ define Device/compex_wpj428
...
+	# wrgg-image.
+	# Factory image size must be larger than 6MB, and size in wrgg header must
+	# match actual factory image size to be flashable from D-Link http server.
+	# Bootloader verifies checksum of wrgg image before booting, thus jffs2
+	# cannot be part of the wrgg image. This is solved in the factory image by
+	# having the rootfs at the end of the image (without pad-rootfs). And in
+	# the sysupgrade image only the kernel is included in the wrgg checksum,
+	# but this is not flashable from the D-link http server.
+	# append-rootfs must start on an erase block boundary.
+	IMAGE/factory.bin    := pad-to 160 | append-kernel | pad-to 6M | append-rootfs | wrgg-image | check-size $$$$(IMAGE_SIZE)
```
Comment (chunkeey) [inline target/linux/ipq40xx/image/Makefile]: please see comment above (pad-to vs pad-extra).

**Comment by chunkeey** [comment]
Please note that the OpenWrt Wiki team might end up reading your commit message (the one with the patch!) in order to fill out the hardware table on the homepage. This is why a feature lists like:

SOC: IPQ4018
DRAM: XYZ MB
ETH: ... (POE?, WAN/LAN?)
LEDS: ...
INPUT: ...
FLASH: ... 

on top the top of the message is great to have. So please take a moment and add these as well.

File: include/image.mk (hunk 5 above)
Comment (frolswe) [reply include/image.mk]: Ahh, right. pad-extra works just as well for me.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 9 above)
Comment (frolswe) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Sure. I was unsure what was the proper OpenWrt way of using the LEDs. Imitating  the factory behaviour works good.

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 4 above)
Comment (mans0n) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: @robimarko A bit off-topic, but could you provide a reference to it? Without knowing it I made a commit 80839694056d raising frequency to 30MHz. If it's true, I guess we'd better to revert the commit.

Comment (robimarko) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Actually, according to the driver comments and defines it appers that QUP controller can go up to 50MHz, anything above 26MHz is defined as high speed in the driver.

Comment (mans0n) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: Oh, so you were talking about this: https://elixir.bootlin.com/linux/latest/source/drivers/spi/spi-qup.c#L119
And that explains why ens620ext used 50MHz... https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts#L167
So frequency above 24MHz is not totally prohibited... Thanks for the information. :)

File: target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts (hunk 11 above)
Comment (frolswe) [reply target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-dlink-dap-2610.dts]: This was some work, but I can see your point. I rebased the DTS, and did not enable the usb devices since they are not actually available on this hardware.

File: package/firmware/ipq-wifi/Makefile (hunk 6 above)
Comment (frolswe) [reply package/firmware/ipq-wifi/Makefile]: I have now B)

**Comment by frolswe** [comment]
The latest version tries to consider @chunkeey comments. Please let me know you want any more changes.

File: include/image-commands.mk [hunk 14]
```diff
@@ -355,6 +355,21 @@ define Build/tplink-v2-image
 	rm -rf $@.new
 endef
 
+# Overwrite the first 160 bytes of the image with wrgg header.
+# Make room for this header by building the image with e.g. "pad-extra 160 | append-kernel | ..."
+define Build/wrgg-image
```
Comment (chunkeey) [inline include/image-commands.mk]: Hm, are you planning on adding more D-Link devices? Or will you convert the existing gemini (DIR-685) and ar71xx (dap-2695-a1) device to this helper? Because it looks a bit out of place since I think, you could have gotten away with just adding this right after "Build/SenaoFW" in the target/linux/ipq40xx/image/Makefile and have somebody else worry about making it all work with the gemini and ar71xx device.

File: include/image-commands.mk [hunk 15]
```diff
@@ -355,6 +355,21 @@ define Build/tplink-v2-image
...
+	dd if=$@ of=$@.new iflag=skip_bytes skip=160
```
Comment (chunkeey) [inline include/image-commands.mk]: This looks like it removes the 160 that got added by pad-extra?! Do you have to do this because you need to have the correct padding for the "pad-to 8M" / "pad-to 64k" down in your image generation rules? If so, then maybe "pad-offset" (see how the the DIR-685 or DAP-2695-A1 does it) is the way to go.

File: target/linux/ipq40xx/image/Makefile [hunk 16]
```diff
@@ -217,6 +217,32 @@ define Device/compex_wpj428
...
 TARGET_DEVICES += compex_wpj428
 
+define Device/dlink_dap-2610
+	$(call Device/FitImageLzma)
+	DEVICE_VENDOR := D-Link
+	DEVICE_MODEL := DAP-2610
+	SOC := qcom-ipq4018
+	DEVICE_DTS_CONFIG := config@ap.dk01.1-c1
+	WRGG_DEVNAME := /dev/mtdblock/8
+	WRGG_SIGNATURE := wapac30_dkbs_dap2610
```
Comment (chunkeey) [inline target/linux/ipq40xx/image/Makefile]: Bit OT: Are the WRGG_VERSION, WRGG_MODEL and WRGG_BUILDNO set somewhere that I can't see?

(If not, you could just drop them from the Build/wrgg-image for now, if you plan on moving the helper to this file anyway).

File: include/image-commands.mk (hunk 14 above)
Comment (frolswe) [reply include/image-commands.mk]: I had some intention of doing that, but I moved the function to the Makefile for now. If I or someone else gets around to unifying the mkwrgg functions, s/he move it again.

File: include/image-commands.mk (hunk 15 above)
Comment (frolswe) [reply include/image-commands.mk]: I originally felt this was a less fragile way than correctly handling the offset in every following step of the image rule. But sure, this works just as well, and with the benefit of being the same as the other firmwares.

File: target/linux/ipq40xx/image/Makefile (hunk 16 above)
Comment (frolswe) [reply target/linux/ipq40xx/image/Makefile]: Ok.

**Comment by chunkeey** [comment]
Uh, I forgot to tell you. I've added the patch to my staging tree (with some changes).

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=41ffbc63e5d3d475c5c5ebad19e22cc781b34f66

There's one small thing with the ethernet vlantag id. (And I think I also need to remove the dangling ethernet1 alias)

**Comment by frolswe** [comment]
That is nice to hear :-)
Thanks to everyone who helped with the review.

**Comment by chunkeey** [comment]
Merged! Thanks.

**Comment by ynezz** [comment]
@frolswe FYI there is a report from one user, that 5.4 kernel bump resulted in the bricked device https://forum.openwrt.org/t/dap-2610-bricked-help-needed, could you check latest snapshot image with 5.4 kernel on your device?
